### PR TITLE
Update mobileAPI.py

### PR DESCRIPTION
**Pull request by shashank136**
Added comments for @gupta-shantanu to review and explain if i am right with understanding of the code.

### CODE REVIEW

File: mobileAPI.py [hunk 1]
```diff
@@ -4,82 +4,127 @@
...
+
+class processMobileSTDcode:
+
+    def processSTD(self,city,n):
+
+        telephone = n
+        ph_len = len(telephone.replace('-',''))
+        # print("length of telephone number is: "+str(ph_len))
+
+        readfile = open('std_data.csv',"r")
```
Comment (rohitmanglik) [inline mobileAPI.py]: Please put this file in ref/pin_code.csv

File: mobileAPI.py [hunk 2]
```diff
@@ -4,82 +4,127 @@
...
-            noslist1=list(set(noslist1)) # creating a list out of noslist1
+            noslist1=list(set(noslist1))
             #print 'noslist1',noslist1
             for i in range(len(noslist1)):
                 # row['Phone'+str((i+1))]=noslist1[i]
-                phn_no_ls+=noslist1[i]+',' # adding phone nubers to string phone_no_ls separated by commas
-            phn_no_ls=phn_no_ls.split(',') # split by comma
-        
-            phn_no_ls=filter(None,phn_no_ls) # saves the filter result in phn_no_ls
+                phn_no_ls+=noslist1[i]+','
```
Comment (rohitmanglik) [inline mobileAPI.py]: why did you remove comments?

File: mobileAPI.py [hunk 3]
```diff
@@ -4,82 +4,127 @@
...
-                    row['Phone'+str(i)]=None  # removed the number from the column
+                    noslist1.append(y)
+                    row['Phone'+str(i)]=None
                     # nphn=row['Phone'+str(i)].replace(' ','').replace('(','').replace(')','').replace('-','')
                     # if len(nphn)>=20:
                     #     phn_no_ls+=
                     # phn_no_ls+=row['Phone'+str(i)]+','
 
             noslist1=filter(None,noslist1);
-            noslist1=list(set(noslist1)) # creating a list out of noslist1
```
Comment (rohitmanglik) [inline mobileAPI.py]: why did you remove comments?

File: mobileAPI.py [hunk 4]
```diff
@@ -4,82 +4,127 @@
...
+        for row in rows:
             noslist=[]
             noslist1=[]
             phn_no_ls=''
             for i in range(1,6):
                 nphn=''
-                if row['Phone'+str(i)] is not None: # getting Phone1 column data from the row
+                if row['Phone'+str(i)] is not None:
                     y=(filter(lambda x: not x.isalpha(),safe_dec_enc(row['Phone'+str(i)])).strip().replace(' ',''))
-                    noslist1.append(y) # inserting that no. into noslist1
```
Comment (rohitmanglik) [inline mobileAPI.py]: why did you remove comments?

File: mobileAPI.py [hunk 5]
```diff
@@ -4,82 +4,127 @@
...
-            return result['national_format'].count(' ') == 1, result['phone_number'] # return national-format and phone number e.g. +91-703xxxxxxx
+            return result['national_format'].count(' ') == 1, result['phone_number']
         except:
-            print "Bad Phone number "+str(phone_no) 
+            print "Bad Phone number "+str(phone_no)
             return False, phone_no
 
-    def _addCountryCode(self,phone_no): # adds or correcting country code like +91
-        phone_no=phone_no.replace(' ','').replace('(','').replace(')','') # removes all white space,(,) from phone number
-        if '0091' in phone_no[0:len(phone_no)/2+1]: # remove 0091 is given as country code
```
Comment (rohitmanglik) [inline mobileAPI.py]: why did you remove comments?

File: mobileAPI.py [hunk 6]
```diff
@@ -4,82 +4,127 @@
...
+        if ph_len < 9:
+            city = ro['City']
+
+            for row in filereader:
+                for field in row:
+                    if field == city.upper():
+                        stdno = eval(row[2])
+                        # print(stdno)
+                        # print(len(str(stdno)))
+                        x=len(str(stdno))
```
Comment (rohitmanglik) [inline mobileAPI.py]: Use better nomenclature.. "x" doesn't mean anything.

File: ref/don'tread.txt [hunk 7]
```diff
@@ -0,0 +1 @@
+
```
Comment (anantharam) [inline ref/don'tread.txt]: Why is this file added to this pull request?

File: mobileAPI.py [hunk 8]
```diff
@@ -4,50 +4,98 @@
...
+
+    def processSTD(self,city,telephone):
+
+        ph_len = len(telephone.replace('-','').repace(" ","")) # remove extra space and invaid characters e.g. '54-66121  3231'
+        # print("length of telephone number is: "+str(ph_len))
+
+        readfile = open('std_data.csv',"r")
+        filereader = csv.reader(readfile)
+
+        if ph_len < 9:
```
Comment (anantharam) [inline mobileAPI.py]: I see even the example phone number(54-66121  3231) is more than 9 digits long, why do we have this check?

File: mobileAPI.py [hunk 9]
```diff
@@ -58,22 +106,32 @@ def checkNo(self,rows):
             phn_no_ls=''
             for i in range(1,6):
                 nphn=''
-                if row['Phone'+str(i)] is not None:
+                if row['Phone'+str(i)] is not None: # getting Phone1 column data from the row
+                    
+                    # Adding STD code to telephone no.
+                    
+                    if len(row['Phone'+str(i)])<9:
```
Comment (anantharam) [inline mobileAPI.py]: Again why do we have this check for less than 9 digits?

File: mobileAPI.py [hunk 10]
```diff
@@ -4,50 +4,98 @@
...
-        global key_index
-        start = key_index
-        for i in range(len(keys)):
+    def _request(self,phone_no): # This function verify if phone number working using Keys 
+        global key_index # key_index =0
+        start = key_index # start = 0
+        for i in range(len(keys)): # iterating through keys 
             try:
-                return json.loads(requests.get(self.url%phone_no,auth=self.key).text)
+                return json.loads(requests.get(self.url%phone_no,auth=self.key).text) # responce in json format
```
Comment (anantharam) [inline mobileAPI.py]: Split this into multiple lines and assign temporary variables for these arguments, they would be very hard to debug otherwise.

File: mobileAPI.py [hunk 11]
```diff
@@ -4,50 +4,98 @@
...
+                        std_len = len(str(stdno))
+                        
+                        if std_len == 2:
+                            std_str = '%0*d' % (3,stdno) # e.g. converts 11 to 011
+                        
+                        elif std_len == 3:
+                            std_str = '%0*d' % (4,stdno) # e.g. converts 524 to 0524
+                        
+                        else:
+                            std_str = '%0*d' % (5,stdno) # e.g. converts 1254 to 01254
```
Comment (anantharam) [inline mobileAPI.py]: If adding zero is just what we want, why this split?
Adding zero at beginning should resolve this without this if statement.

File: mobileAPI.py (hunk 8 above)
Comment (shashank136) [inline mobileAPI.py]: sir, on line no.17 `ph_len = len(telephone.replace('-','').repace(" ","")) ` we are removing the extra spaces and '-' to get an ideal phone number without any invalid characters.
Like "54-66121 3231" will change into 54661213231 which is actually an invalid number thus we check if the number is less then 9 digits before preceding to add STD code to telephone numbers (like to 270xxx) is a valid telephone number to which STD code will be added.

File: mobileAPI.py (hunk 11 above)
Comment (shashank136) [inline mobileAPI.py]: will update that section of code

File: mobileAPI.py (hunk 9 above)
Comment (shashank136) [inline mobileAPI.py]: will optimize this section.

Comment (shashank136) [inline mobileAPI.py]: sir can u check the unit test code in test folder. Rohan sir asked to modify it to use the sample_data.csv file....if it meets the requirement then we can precede run the test for large set of sample data available in input directory. @anantharam.

File: test/test_updatedprocessMobile.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+	src_list = []
+
+	for row in src_reader:
+		x = row
+		src_list.append(x)
+
+	val = False
+	val = updatedprocessMobile.processAll(src_list)
+	assert val== True
```
Comment (anantharam) [inline test/test_updatedprocessMobile.py]: What exactly are you asserting in the unit test?

File: mobileAPI.py (hunk 9 above)
Comment (anantharam) [inline mobileAPI.py]: Is the unit test completed? I don't see anything getting asserted in the unit test.

File: mobileAPI.py (hunk 8 above)
Comment (anantharam) [inline mobileAPI.py]: Can you also add an example for a valid number in the comments?

File: mobileAPI.py (hunk 9 above)
Comment (shashank136) [inline mobileAPI.py]: Sir, i checked for a very small sample, so we can say it is not tested for real time dataset. Here, we are trying to assert if we are successfully enter the corrected phone number into final 'Mobile' field in the csv. If we are able to add the updated phone number to mobile field then it return TRUE to test case else false (Here corrected phone number is mobile number in international format accepted and for the telephone numbers with STD code in international format).

File: test/test_updatedprocessMobile.py (hunk 12 above)
Comment (shashank136) [inline test/test_updatedprocessMobile.py]: Sir, i checked for a very small sample, so we can say it is not tested for real time dataset (samples available in input directory). Here, we are trying to assert if we are successfully enter the corrected phone number into final 'Mobile' field in the csv. If we are able to add the updated phone number to mobile field then it return TRUE to test case else false (Here corrected phone number is mobile number in international format accepted and for the telephone numbers with STD code in international format).

Comment (anantharam) [inline test/test_updatedprocessMobile.py]: As discussed, please update the test cases to check for failure cases too, not just one single test case.

File: test/test_updatedprocessMobile.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+		src_list.append(x)
+
+	# val = False
+	tabb = updatedprocessMobile.processAll(src_list)
+
+	 for i in tabb:
+		if tabb[i] == True:
+			print(str(i)+' '+'following is a vaid number')
+		else:
+			print(str(i)+' '+'following is an invalid number')
```
Comment (anantharam) [inline test/test_updatedprocessMobile.py]: Printing the result is not a test case, check if the expected result is true and it is actually true
